language/move-stdlib/nursery/Vault.move

```diff
@@ -0,0 +1,462 @@
+    /// should probably be relaxed in the languages, and only types actually used in the data should influence the
+    /// ability requirements.
+    struct ReadCap<Content: store + drop> has copy, drop {
+        vault_address: address,
+        authority: address
+    }
+
+    /// A capability to modify the content of the vault.
+    struct ModifyCap<Content: store + drop> has copy, drop {
+        vault_address: address,
+        authority: address
+    }
+
+    /// A capability to delegate access to the vault.
+    struct DelegateCap<Content: store + drop> has copy, drop {
+        vault_address: address,
+        authority: address
+    }
+
+    /// A capability to transfer ownership of the vault.
+    struct TransferCap<Content: store + drop> has copy, drop {
+        vault_address: address,
+        authority: address
+    }
+
+    /// A type describing a capability. This is used for functions like `Self::delegate` where we need to
+    /// specify capability types.
+    struct CapType has copy, drop, store { code: u8 }
+
+    /// Creates a read capability type.
+    public fun read_cap_type(): CapType { CapType{ code : 0 } }
+
+    /// Creates a modify  capability type.
+    public fun modify_cap_type(): CapType { CapType{ code : 1 } }
+
+    /// Creates a delegate  capability type.
+    public fun delegate_cap_type(): CapType { CapType{ code : 2 } }
+
+    /// Creates a transfer  capability type.
+    public fun transfer_cap_type(): CapType { CapType{ code : 3 } }
+
+    // ================================================================================================================
+    // Events
+
+    /// An event which we generate on vault access delegation or revocation.
+    struct VaultDelegateEvent has store, drop {
+        metadata: vector<u8>,
+        vault_address: address,
+        authority: address,
+        delegate: address,
+        cap: CapType,
+        is_revoked: bool,
+    }
+
+    /// An event which we generate on vault transfer.
+    struct VaultTransferEvent has store, drop {
+        metadata: vector<u8>,
+        vault_address: address,
+        authority: address,
+        new_vault_address: address,
+    }
+
+    // ================================================================================================================
+    // Representation
+
+    /// Private. The vault representation.
+    struct Vault<Content: store + drop> has key {
+        /// Contains any metadata which identifies this vault. This information is used
+        /// in events generated by this module.
+        metadata: vector<u8>,
+        /// Whether this vault supports delegation.
+        supports_delegation: bool,
+        /// Event handle for vault delegation.
+        delegate_events: Event::EventHandle<VaultDelegateEvent>,
+        /// Event handle for vault transfer.
+        transfer_events: Event::EventHandle<VaultTransferEvent>,
```

Comment:
It feels a bit weird to me that this is a part of this system and not something handled by someone who is using the vault.

Reply:
I was debating about this as well and whether these events should be changed when a Vault is transferred. I actually think the fixity of this event handle could be quite useful -- if you have an interest in a particular `Vault<T>` you can just listen on one event handle to figure out who the current owner is (and then use this to e.g., request access and/or read the data off-chain).

Reply:
Given the way this is setup, you could split out the event handles into a separate resource. And sort of make them opt-in (or just an option in this resource)

Reply:
I like the idea of making this optional, but if we want to come with a design which prescribes a good design pattern for secure storage, it should be there, and it should be on by default. If someone doesn't want the feature of auditing important security event. he should explicitly opt out.

+1 to moving in a different resource if it is optional.